src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs

```diff
@@ -255,6 +257,39 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
         appModel.Resources.AddRange(sideCars);
     }
 
+    // This method resolves the application's endpoint and the protocol that the dapr side car will use.
+    // It depends on DaprSidecarOptions.AppProtocol and DaprSidecarOptions.AppEndpoint.
+    // - If both are null default to 'http' for both.
+    // - If AppProtocol is not null try to get an endpoint with the name of the protocol.
+    // - if AppEndpoint is not null try to use the scheme as the protocol.
+    // - if both are not null just use both options.
+    static (EndpointReference appEndpoint, string protocol)? GetEndpointReference(DaprSidecarOptions? sidecarOptions, IResource resource)
+    {
+        if (resource is IResourceWithEndpoints resourceWithEndpoints)
+        {
+            if (sidecarOptions == null || (sidecarOptions.AppProtocol == null && sidecarOptions.AppEndpoint == null))
```

Comment:
Can you rewrite this as a switch expression?

Reply:
@davidfowl Do you have time for a bit of guidance here.
I tried it but failed and I don't know what I do wrong:

```csharp
return (sidecarOptions, resourceWithEndpoints) switch{
                (DaprSidecarOptions? p0, IResourceWithEndpoints p1) when p0 == null || (p0.AppProtocol == null && p0.AppEndpoint == null) => (p1.GetEndpoint("http"), "http"),
                _ => throw new ArgumentException($"The combination of {nameof(DaprSidecarOptions.AppEndpoint)} and {nameof(DaprSidecarOptions.AppProtocol)}")
            }
```

That's my current code. And it will result in:

"CS0119", 'DaprSidecarOptions' is a type, which is not valid in the given context"

Reply:
@davidfowl nevermind... I just used `var` and it works...

Reply:
I think it would look something like this:

```C#
if (resource is not IResourceWithEndpoints resourceWithEndpoints)
{
    return null;
}

return (sidecarOptions?.AppProtocol, sidecarOptions?.AppEndpoint) switch
{
    // If both are not null, use AppEndpoint and AppProtocol
    (string appProtocol, string appEndpoint) => (resourceWithEndpoints.GetEndpoint(appEndpoint), appProtocol),

    // If both are null, default to http
    (null, null) => (resourceWithEndpoints.GetEndpoint("http"), "http"),

    // If AppProtocol is null, use AppEndpoint
    (null, string appEndpoint) => (resourceWithEndpoints.GetEndpoint(appEndpoint), appEndpoint),

    // If both are not null, use AppEndpoint and AppProtocol
    (string appProtocol, null) => (resourceWithEndpoints.GetEndpoint(appProtocol), appProtocol);
}
```